base/applications/cmdutils/taskkill/taskkill.c

```diff
@@ -237,69 +289,116 @@ static int send_close_messages(void)
-                    taskkill_message_printfW(STRING_CLOSE_PROC_SRCH, process_name, pid_list[index]);
+                    pkill_list[pkill_size] = pid_list[index];
+                    pkill_size++;
                 }
             }
+        }
+
+        // Can't find any process same as name or PID
+        if (pkill_size == 0)
+        {
+            taskkill_message_printfW(STRING_SEARCH_FAILED, task_list[i]);
+            status_code = 128;
+            continue;
+        }
+
+        // Try to send close messages to process in `pkill_list`
+        for (DWORD index = 0; index < pkill_size; index++)
+        {
+            DWORD pid = pkill_list[index];
+            WCHAR process_name[MAX_PATH] = { 0 };
+            struct pid_close_info info = { pid };
 
-            if (!found_process)
+            if (pid == self_pid)
             {
-                taskkill_message_printfW(STRING_SEARCH_FAILED, task_list[i]);
-                status_code = 128;
+                taskkill_message(STRING_SELF_TERMINATION);
+                status_code = 1;
+                continue;
+            }
+
+            // Send close messages to child first
+            if (kill_child_processes)
+            {
+                send_close_messages_tree(pid);
+            }
+
+            get_process_name_from_pid(pid, process_name, MAX_PATH);
+            EnumWindows(pid_enum_proc, (LPARAM)&info);
+            if (info.found)
+            {
+                if (is_numeric)
+                {
+                    taskkill_message_printfW(STRING_CLOSE_PID_SEARCH, pid);
+                }
+                else
+                {
+                    taskkill_message_printfW(STRING_CLOSE_PROC_SRCH, process_name, pid);
+                }
             }
         }
+
+        HeapFree(GetProcessHeap(), 0, pkill_list);
     }
 
     HeapFree(GetProcessHeap(), 0, pid_list);
     return status_code;
 }
 
+static void terminate_process_tree(DWORD ppid)
+{
+    DWORD child_size = 0, index = 0;
+    DWORD* child_pid_list = find_child_pid(ppid, &child_size);
+
+    // Use recursion to browse all child processes
+    while (index < child_size)
+    {
+        terminate_process_tree(child_pid_list[index]);
+        HANDLE process = OpenProcess(PROCESS_TERMINATE, FALSE, child_pid_list[index]);
+        if (!process)
+        {
+            index++;
+            continue;
+        }
+
+        if (!TerminateProcess(process, 0))
+        {
+            taskkill_message_printfW(STRING_TERM_CHILD_FAILED, child_pid_list[index], ppid);
+            CloseHandle(process);
+            index++;
+            continue;
+        }
+
+        taskkill_message_printfW(STRING_TERM_CHILD, child_pid_list[index], ppid);
+        CloseHandle(process);
+        index++;
+    }
+
+    HeapFree(GetProcessHeap(), 0, child_pid_list);
+}
+
 static int terminate_processes(void)
 {
-    DWORD *pid_list, pid_list_size;
+    DWORD* pid_list, pid_list_size;
```

Comment:
Is this based on anything other than personal preference? Space before asterisk seems to be the prevalent style in this file so should be adhered to.

Reply:
I don't know why I do that :p I fixed. Thanks for your recommend.